### ISSUE TITLE
OrderManipulation updateActionsForm doesn't work

### ISSUE THREAD

**Issue by a2nt** [silvershop/silvershop-core#538]
All those ways doesn't work for some reason:
```
AccountPage_Controller:
  extensions:
    - AccountPage_ControllerExtension
```
```
OrderManipulation:
  extensions:
    - OrderManipulationExtension
```
```
OrderActionsForm:
  extensions:
    - OrderActionsFormExtension
```

**Comment by bummzack** [comment]
I don't like the `updateActionsForm` extension point. It's being called from an extension and applied to a form… that's really confusing.
I think it would be safe to remove `updateActionsForm`, since you can just use `updateForm` on an extension applied to `OrderActionsForm`.

Eg. 
```yaml
OrderActionsForm:
  extensions:
    - OrderActionsFormExtension
```

And then in `OrderActionsFormExtension.php`:

```php
public function updateForm($order)
{
    // Do something with $this->owner, as it's the Form instance
}
```

**Comment by a2nt** [comment]
Cool can we make it work with the other forms?

**Comment by bummzack** [comment]
What other forms are you talking about? This issue is about `OrderManipulation` and `updateActionsForm`, or am I missing something?

**Comment by a2nt** [comment]
Well here's an example extension:
```
class ShopBootstrapFormEx extends Extension {
    public function updateForm(){
        $fields = $this->owner->Fields();
        BootstrapForm::apply_bootstrap_to_fieldlist($fields);
        $fields = $this->owner->Actions();
        BootstrapForm::apply_bootstrap_to_fieldlist($fields);
        $this->owner->addExtraClass('form form-vertical');
    }
}
```

It will work with OrderActionsForm by:
```
OrderActionsForm:
  extensions:
    - ShopBootstrapFormEx
```
but won't work with ShopAccountForm cuz it requires updateShopAccountForm() method it will be more clear to use same methods to do the same things

**Comment by bummzack** [comment]
Yeah, you're right about that. It's not really consistent… If you want to unify all the extension-points, a PR would be welcome. We'd have to keep the old ones intact though and add deprecation warnings for them (we can remove them as early as version `3.0`).

As a workaround, you could just add `updateShopAccountForm` to your extension and implement it like this:

```php
public function updateShopAccountForm()
{
    $this->updateForm();
}
```

**Comment by a2nt** [comment]
That's what I do, but it's kinda dirty way that's why I offer to standardize this methods. Keeping backwards compatibility of course which can be removed in future.

**Comment by bummzack** [comment]
An argument against a unified extension-point is, that you'll have to create custom extensions for *every* form you want to treat differently.

So without the unified extension-points you could create an extension called `ShopFormsExtension`, like this:

```php
class ShopFormsExtension extends DataExtension
{
	public function updateShopAccountForm()
	{
		// do some specific stuff to ShopAccountForm
	}

	public function updateChangePasswordForm()
	{
		// something specific to ChangePasswordForm
	}

	// … etc …
}
```

And then you could just apply it to all the forms and still have the possibility to customize every form individually.

So the logical follow up to this approach would be **not** to name all extension points `updateForm`, but rather rename `updateForm` to `updateOrderActionsForm`…

I think this would actually be more in line with what they do in the SilverStripe Framework, where there are extensions like `CMSMain.updateEditForm`, `LeftAndMain.updateBatchActionsForm` etc.

**Comment by a2nt** [comment]
well to treat forms differently I can create different extensions

**Comment by bummzack** [comment]
I know… but you could also argue that this is cumbersome and it's easier to just have different methods instead? 

Personally, I'm more inclined to go the same route as SilverStripe core does, eg. having different methods for different forms. But I'd like to get some more opinions on this. Pinging @anselmdk, @wilr, @chillu

**Comment by a2nt** [comment]
well in this case let's add all methods directly to form objects. Because for an instance PaymentForm can't be extended this way and you need to extend CheckoutPage and every Checkout step.

**Comment by a2nt** [comment]
I think that it will solve everything with having "updateForm" methods for Form objects and "updateShopAccountForm" like methods for a specific places at Controllers. What do you think?

**Comment by bummzack** [comment]
I agree that it's currently a bit of a mess. From what I can see in SilverStripe core, there's usually an extension hook in the method that *creates* the form, and the form is then passed to the extension as parameter. Eg. `$this->extend('updateEditForm', $form);`

I guess some of the extension hooks were put inside the form constructor, because the forms were instantiated from different places? @markguinn any insights on this?

There are several possible scenarios:

 1. **All** custom forms within shop have an extension hook `updateForm` within their constructor. Non-custom forms (an example would be `AccountPage_Controller.DefaultAddressForm` which returns a normal `Form` instance) use the `$this->extend('updateFormName', $form);` pattern.
 2. Same as above, but instead of `updateForm` also use the `updateFormName` pattern for the custom forms. Eg. within the `PaymentForm` constructor, use `$this->extend('updatePaymentForm');`.
 3. Get rid of the extension points inside of form constructors and ensure that  `$this->extend('updateFormName', $form);` is used consistently wherever a form instance will be created/returned.

My vote would go to solution `2`, since it's closest to what we currently have **and** I prefer having multiple methods in one multipurpose extension over having to create multiple extensions.

**Comment by a2nt** [comment]
In case of option (2) it can be messy on child objects.
For an instance PaymentForm extended from CheckoutForm if you will use $this->extend('updateFormName', $form); you will need to repeat yourself on PaymentForm too.

**Comment by bummzack** [comment]
That would be option 3, where we remove the extension point from the form.
Options 1 + 2 will have the extension point within the form constructor for all custom forms and **only** use `$this->extend('updateFormName', $form);` where non-custom Form instances are created…

The only difference between option 1 and 2 is that the latter will still have different names for the extension-hooks instead of using always the same name for the hook.

**Comment by a2nt** [comment]
that's sounds reasonable there will be no repeated hooks so option (3) I think is the best, but in this way we shouldn't get rid of updateShopAccountForm method
